### PR TITLE
dsa 4

### PR DESCRIPTION
**Pull request by meta-dipanshu-sharma**

### CODE REVIEW

File: PF/DSA-4/q1_employee/Employee.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        return address;
+    }
+    
+    /**
+     * Compares employee id
+     */
+    @Override
+    public int compareTo(Employee employee)
+    {
```
Comment (chirag-metacube) [inline PF/DSA-4/q1_employee/Employee.java]: This method can be return id - employee.getId()

File: PF/DSA-4/q1_employee/EmployeeOperations.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    }
+     
+    /**
+     * Returns list of employee sorted by id
+     * @return list of employee
+     */
+    public List<Employee> sortById()
+    {
+        List<Employee> sortedEmployeeList = employeeList;
+        Collections.sort(sortedEmployeeList);
```
Comment (chirag-metacube) [inline PF/DSA-4/q1_employee/EmployeeOperations.java]: return Collections.sort(sortedEmployeeList);

File: PF/DSA-4/q1_employee/EmployeeOperationsTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        boolean actualOutput = employeeOperations.addEmployee(2, "sehwag", "delhi");
+        assertEquals(expectedOutput, actualOutput);
+    }
+
+    @Test
+    public void givenSortEmployee_whenSortingIsDoneByName_thenCodeRunsSuccessfully() {
+        String[] expectedOutput = { "chahal", "dhoni", "dipanshu", "virat", "zaheer" };
+        int i = 0;
+        for (Employee employee : employeeOperations.sortByName()) {
+            assertEquals(expectedOutput[i], employee.getName());
```
Comment (chirag-metacube) [inline PF/DSA-4/q1_employee/EmployeeOperationsTest.java]: Wrong way to assert list

File: PF/DSA-4/q1_employee/SortByName.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+{
+
+    /**
+     * This method overrides compare method of Comparator and is used for comparing two strings lexicographically
+     * @return +ve value if employee1 name comes first in dictionary and -ve if employee2 comes first and 0 if both are equal.
+     * */
+    @Override
+    public int compare(Employee employee1, Employee employee2)
+    {
+
```
Comment (chirag-metacube) [inline PF/DSA-4/q1_employee/SortByName.java]: extra line

File: PF/DSA-4/q2_uniquecharacters/UniqueCharacters.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+     * @param inputString string
+     * @return number of unique characters in the string
+     */
+    public int findNumberOfUniqueCharacters(String inputString)
+    {
+        Set<Character> characters = new HashSet<Character>();
+        Set<Character> duplicateCharacters = new HashSet<Character>();
+        if(previouslyEncounteredString.containsKey(inputString))
+        {
+            //System.out.println("From map");
```
Comment (chirag-metacube) [inline PF/DSA-4/q2_uniquecharacters/UniqueCharacters.java]: remove this

File: PF/DSA-4/q2_uniquecharacters/UniqueCharacters.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    
+    /**
+     * Finds count of unique characters in the string
+     * @param inputString string
+     * @return number of unique characters in the string
+     */
+    public int findNumberOfUniqueCharacters(String inputString)
+    {
+        Set<Character> characters = new HashSet<Character>();
+        Set<Character> duplicateCharacters = new HashSet<Character>();
```
Comment (chirag-metacube) [inline PF/DSA-4/q2_uniquecharacters/UniqueCharacters.java]: No need of this

File: PF/DSA-4/q2_uniquecharacters/UniqueCharacters.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+ /**
+ * @author Dipanshu sharma
+ * This is class which have method to count no of unique characters in a string.
+ */
+public class UniqueCharacters
+{
+    private Map<String, Integer> previouslyEncounteredString;
```
Comment (chirag-metacube) [inline PF/DSA-4/q2_uniquecharacters/UniqueCharacters.java]: Name can be "cache"

File: PF/DSA-4/q2_uniquecharacters/UniqueCharacters.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+            for(int i=0; i<inputString.length(); i++)
+            {
+                if(duplicateCharacters.contains(inputString.charAt(i)))
+                {
+                    continue;
+                }
+                if(characters.contains(inputString.charAt(i)))
+                {
+                    characters.remove(inputString.charAt(i));
+                    duplicateCharacters.add(inputString.charAt(i));
```
Comment (chirag-metacube) [inline PF/DSA-4/q2_uniquecharacters/UniqueCharacters.java]: Set functions can directly be used

File: PF/DSA-4/q2_uniquecharacters/UniqueCharactersTest.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    }
+    
+    @Test
+    public void givenStringOccurredPreviously_whenToFindUniqueCharacters_thenCodeRunsSuccessfully()
+    {
+        int expectedOutput = 8;
+        int actualOutput = uniqueCharacters.findNumberOfUniqueCharacters("dipanshu");
+        assertEquals(expectedOutput, actualOutput);
+    }
+
```
Comment (chirag-metacube) [inline PF/DSA-4/q2_uniquecharacters/UniqueCharactersTest.java]: Boundary case test cases missing
